core/server/common/src/main/java/alluxio/master/journal/AsyncJournalWriter.java

```diff
@@ -99,49 +146,161 @@ public long appendEntry(JournalEntry entry) {
+        while (!mQueue.isEmpty()) {
           // Get, but do not remove, the head entry.
           JournalEntry entry = mQueue.peek();
           if (entry == null) {
-            // No more entries in the queue. Break out of the infinite for-loop.
+            // No more entries in the queue. Break write session.
             break;
           }
           mJournalWriter.write(entry);
+          needFlush = true;
           // Remove the head entry, after the entry was successfully written.
           mQueue.poll();
           writeCounter = mWriteCounter.incrementAndGet();
 
-          if (writeCounter >= targetCounter) {
-            if ((System.nanoTime() - startTime) >= mFlushBatchTimeNs) {
-              // This thread has been writing to the journal for enough time. Break out of the
-              // infinite for-loop.
-              break;
+          if ((System.nanoTime() - startTime) >= mFlushBatchTimeNs) {
+            // This thread has been writing to the journal for enough time. Break out of the
+            // infinite while-loop.
+            break;
+          }
+        }
+
+        if (needFlush) {
+          mJournalWriter.flush();
+          mFlushCounter.set(writeCounter);
+
+          /**
+           * Notify tickets that have been served to wake up.
+           */
+          List<FlushTicket> closedTickets = new ArrayList<>();
+          try (LockResource lr = new LockResource(mTicketLock)) {
+            for (FlushTicket ticket : mTicketList) {
+              if (ticket.getTargetCounter() <= writeCounter) {
+                ticket.setCompleted();
+                closedTickets.add(ticket);
+              }
             }
+            mTicketList.removeAll(closedTickets);
           }
         }
+      } catch (IOException | JournalClosedException exc) {
+        /**
+         * Release only tickets that have been flushed. Fail the rest.
+         */
+        List<FlushTicket> closedTickets = new ArrayList<>();
+        try (LockResource lr = new LockResource(mTicketLock)) {
+          for (FlushTicket ticket : mTicketList) {
+            closedTickets.add(ticket);
+            if (ticket.getTargetCounter() <= mFlushCounter.get()) {
+              ticket.setCompleted();
+            } else {
+              ticket.setError(exc);
+            }
+          }
+          mTicketList.removeAll(closedTickets);
+        }
+      }
+    }
+  }
+
+  /**
+   * Submits a ticket to flush thread and waits until ticket is served.
+   *
+   * If the specified counter is already flushed, this is essentially a no-op.
+   *
+   * @param targetCounter the counter to flush
+   */
+  @SuppressWarnings("Duplicates")
+  public void flush(final long targetCounter) throws IOException, JournalClosedException {
+    // Return if flushed.
+    if (targetCounter <= mFlushCounter.get()) {
+      return;
+    }
+
+    /**
+     * Submit the ticket for flush thread to process.
+     */
+    FlushTicket ticket = new FlushTicket(targetCounter);
```

Comment:
Could this repeated creation cause a bottleneck?

Reply:
Honestly I don't know. It was not the case hundreds of threads. I don't think it'll cause a bottleneck until a 1000 mark.